### PR TITLE
feat(#4407): add a log histogram 

### PR DESCRIPTION
**Pull request by aptlin**
## Summary

Screenshot 2023-03-07 at 5 21 52 PM

## How did you test this change?

local click test after
```
doppler run -- go run backend/clickhouse/seeds/main.go
```
## Are there any deployment considerations?

no

### CODE REVIEW

File: frontend/src/pages/LogsPage/SearchForm/SearchForm.tsx [hunk 1]
```diff
@@ -195,34 +228,39 @@ const Search: React.FC<{
 		>
 			<IconSolidSearch className={styles.searchIcon} />
 
-			<Combobox
-				autoSelect={autoSelect}
```
Comment (aptlin) [inline frontend/src/pages/LogsPage/SearchForm/SearchForm.tsx:199]: Encountered infinite rerenders when using autoSelect with `formState.setValue` in `onChange`

File: frontend/src/pages/LogsPage/SearchForm/SearchForm.tsx [hunk 2]
```diff
@@ -195,34 +232,39 @@ const Search: React.FC<{
...
+					placeholder="Search your logs..."
+					className={styles.combobox}
+					onBlur={() => {
+						formState.setValue('query', state.value)
+						formState.submit()
+						inputRef?.current?.blur()
+					}}
+				/>
+
+				{isDirty ? (
```
Comment (et) [inline frontend/src/pages/LogsPage/SearchForm/SearchForm.tsx:257]: Nice!

File: frontend/src/pages/LogsPage/SearchForm/SearchForm.tsx [hunk 3]
```diff
@@ -195,34 +232,39 @@ const Search: React.FC<{
...
 
-			<Combobox
-				autoSelect={autoSelect}
-				ref={inputRef}
-				state={state}
-				name="search"
-				placeholder="Search your logs..."
-				value={query}
-				onChange={(e) => {
-					const value = e.target.value
```
Comment (et) [inline frontend/src/pages/LogsPage/SearchForm/SearchForm.tsx:206]: I don't know where but I'm seeing a regression whereby clicking a key doesn't prepopulate until after a value is selected:

Kapture 2023-03-07 at 16 03 07

**Review by et — COMMENTED** [review]
Looking good! Can you check on the regression (log key not being prepopulated after selection)? Hopefully this is a quick fix.

File: frontend/src/pages/LogsPage/LogsPage.tsx [hunk 4]
```diff
@@ -79,30 +58,24 @@ const LogsPage = () => {
...
-						end_date: moment(endDate).format(FORMAT),
-					},
-				},
-			},
-			skip: !project_id,
-		})
-
 	const handleFormSubmit = (value: string) => {
-		setQuery(value)
+		if (!!value) {
```
[the remark this answers is not in the document]
Comment (aptlin) [reply frontend/src/pages/LogsPage/LogsPage.tsx:62]: Nit, but this makes sure that the query url param gets removed if it's empty

File: frontend/src/pages/LogsPage/LogsPage.tsx [hunk 5]
```diff
@@ -79,30 +58,24 @@ const LogsPage = () => {
...
+			setQuery(value)
+		} else {
+			setQuery(undefined)
+		}
 	}
 
 	const handleDatesChange = (newStartDate: Date, newEndDate: Date) => {
+		if (!isSignificantDateRange(newStartDate, newEndDate)) return
```
[the remark this answers is not in the document]
Comment (aptlin) [reply frontend/src/pages/LogsPage/LogsPage.tsx:70]: This avoids confusing UX: if you click on a bucket over a narrow time range or drag select a narrow interval, we get to the point when the startDate equals endDate, which means that no meaningful histogram can be drawn, but some log lines still match that timestamp.

File: frontend/src/pages/LogsPage/LogsHistogram/LogsHistogram.tsx [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+
+	return (
+		<Popover placement="bottom-start" open={open} setOpen={setOpen}>
+			{trigger}
+			{content}
+		</Popover>
+	)
+}
+
+const severityToColor = {
```
[the remark this answers is not in the document]
Comment (aptlin) [reply frontend/src/pages/LogsPage/LogsHistogram/LogsHistogram.tsx]: Great idea! [d78614a](https://github.com/highlight/highlight/pull/4524/commits/d78614a892844183b108bbef5e9ce9416e4d8dbb)

File: frontend/src/pages/LogsPage/LogsPage.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (et) [reply frontend/src/pages/LogsPage/LogsPage.tsx:70]: Ah, I see. Pretty sure I hit this scenario where by the calendar is open and they only choose the start date.

Is it possible to move the logic inside this `useEffect` here?
https://github.com/highlight/highlight/blob/d78614a892844183b108bbef5e9ce9416e4d8dbb/frontend/src/pages/LogsPage/SearchForm/SearchForm.tsx#L59-L68

That way we can consolidate this in one spot.

File: frontend/src/pages/LogsPage/SearchForm/SearchForm.tsx (hunk 3 above)
Comment (aptlin) [reply frontend/src/pages/LogsPage/SearchForm/SearchForm.tsx:206]: Thanks for catching this! Fixed in [f90790c](https://github.com/highlight/highlight/pull/4524/commits/f90790c5e0273bfa568b0ea09de0142305536eb9)

File: frontend/src/pages/LogsPage/LogsPage.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (aptlin) [reply frontend/src/pages/LogsPage/LogsPage.tsx]: Thanks! Done in [e6c450f](https://github.com/highlight/highlight/pull/4524/commits/e6c450f29258dba117c517504090a2f387f4fbcf)

**Pull request merged**